File: src/portfolio_manager/services/stock_service.py
"""Service for resolving and persisting stock display names."""

import logging

from portfolio_manager.services.stock_name_formatter import format_stock_name

logger = logging.getLogger(__name__)


class StockService:
    def __init__(self, stock_repository, price_service=None):
        self._stock_repository = stock_repository
        self._price_service = price_service

    def resolve_and_persist_name(self, stock) -> str:
        """Return formatted display name, persisting it when newly resolved.

        Returns the formatted existing name if already set.
        Calls price_service to look up the name when stock.name is empty.
        Returns "" if no name can be resolved or price_service is unavailable.
        Mutates stock.name in-place when a new name is persisted.
        """
        if stock.name:
            return format_stock_name(stock.name)

        if self._price_service is None:
            return ""

        resolved_name = ""
        try:
            _, _, resolved_name, _ = self._price_service.get_stock_price(
                stock.ticker, preferred_exchange=stock.exchange
            )
        except Exception:
            logger.warning(
                "price_service.get_stock_price failed for %s",
                stock.ticker,
                exc_info=True,
            )
            return ""

        if not resolved_name:
            return ""

        formatted = format_stock_name(resolved_name)
        if not formatted:
            return ""

        self._stock_repository.update_name(stock.id, formatted)
        stock.name = formatted
        return formatted

File: internal/services/stock_service.go
package services

import (
	"context"
	"log"

	"github.com/kadragon/portfolio-manager/internal/models"
	"github.com/kadragon/portfolio-manager/internal/repositories"
	"github.com/kadragon/portfolio-manager/internal/stockformat"
)

// StockService resolves and persists stock display names.
// Mirrors Python's StockService.
type StockService struct {
	stocks       *repositories.StockRepository
	priceService *PriceService
}

// NewStockService creates a StockService. priceService may be nil.
func NewStockService(stocks *repositories.StockRepository, priceService *PriceService) *StockService {
	return &StockService{stocks: stocks, priceService: priceService}
}

// SetPriceService wires in a PriceService after construction (late-binding).
func (s *StockService) SetPriceService(ps *PriceService) {
	s.priceService = ps
}

// PersistName formats raw_name and persists it to the DB when stock.Name is empty.
// Returns the formatted raw_name if non-empty, else the formatted stored name.
func (s *StockService) PersistName(ctx context.Context, stock *models.Stock, rawName string) string {
	formatted := stockformat.FormatName(rawName)
	if stock.Name == "" && formatted != "" {
		updated, err := s.stocks.UpdateName(ctx, stock.ID, formatted)
		if err != nil {
			log.Printf("StockService: persist name for %s: %v", stock.Ticker, err)
		} else {
			stock.Name = updated.Name
			stock.UpdatedAt = updated.UpdatedAt
		}
	}
	if formatted != "" {
		return formatted
	}
	return stockformat.FormatName(stock.Name)
}

// ResolveAndPersistName returns the formatted display name, persisting it when newly resolved.
// Returns the formatted existing name if already set.
// Calls priceService when stock.Name is empty; returns "" if unavailable.
func (s *StockService) ResolveAndPersistName(ctx context.Context, stock *models.Stock) string {
	if stock.Name != "" {
		return stockformat.FormatName(stock.Name)
	}
	if s.priceService == nil {
		return ""
	}

	preferredExchange := ""
	if stock.Exchange != nil {
		preferredExchange = *stock.Exchange
	}
	_, _, resolvedName, _ := s.priceService.GetStockPrice(ctx, stock.Ticker, preferredExchange)
	return s.PersistName(ctx, stock, resolvedName)
}
